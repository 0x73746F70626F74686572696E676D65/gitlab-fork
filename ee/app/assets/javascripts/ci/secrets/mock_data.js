const mockSecretsData = [
  {
    name: 'runner-token-windows',
    labels: [],
    lastAccessed: '2024-03-19T20:55:08.551Z',
    createdAt: '2024-03-11T01:33:06.258Z',
  },
  {
    name: 'product/client-tokens',
    labels: ['env::staging', 'env::production'],
    lastAccessed: '2024-01-23T07:12:44.833Z',
    createdAt: '2024-01-22T08:04:26.024Z',
  },
  {
    name: 'product/server-tokens',
    labels: ['env::staging', 'env::production'],
    lastAccessed: '2024-01-22T07:12:44.833Z',
    createdAt: '2023-01-24T08:04:26.024Z',
  },
  {
    name: 'security-token-project2',
    labels: ['env::production', 'security'],
    lastAccessed: '2024-01-21T23:13:37.227Z',
    createdAt: '2024-01-19T11:59:43.516Z',
  },
  {
    name: 'personal-tokens',
    labels: [],
    lastAccessed: '2024-01-19T21:01:38.925Z',
    createdAt: '2024-01-17T05:51:54.602Z',
  },
  {
    name: 'token-deployment',
    labels: [],
    lastAccessed: '2024-01-16T23:13:31.548Z',
    createdAt: '2024-01-15T17:12:33.882Z',
  },
  {
    name: 'runner-token-mac',
    labels: [],
    lastAccessed: '2024-01-06T07:50:58.308Z',
    createdAt: '2023-01-15T16:35:38.326Z',
  },
  {
    name: 'aws-db-prod-credentials',
    labels: ['AWS'],
    lastAccessed: '2024-01-05T11:47:59.454Z',
    createdAt: '2024-01-04T10:37:41.135Z',
  },
  {
    name: 'runner-token-linux',
    labels: [],
    lastAccessed: '2023-12-30T16:41:08.551Z',
    createdAt: '2023-01-14T18:52:56.258Z',
  },
  {
    name: 'SSH-KEY-1',
    labels: [],
    lastAccessed: '2023-11-29T22:18:31.562Z',
    createdAt: '2023-11-30T01:29:14.394Z',
  },
  {
    name: 'group-ID',
    labels: ['shell'],
    lastAccessed: '2023-11-29T04:30:10.220Z',
    createdAt: '2023-01-21T17:34:33.544Z',
  },
  {
    name: 'security-token-project1',
    labels: ['security'],
    lastAccessed: '2023-11-28T04:30:10.220Z',
    createdAt: '2023-01-21T17:34:33.544Z',
  },
];

export const mockGroupSecretsData = mockSecretsData.map((secret, index) => ({
  key: `group_secret_${index}`,
  ...secret,
}));

export const mockProjectSecretsData = mockSecretsData.map((secret, index) => ({
  key: `project_secret_${index}`,
  ...secret,
}));
