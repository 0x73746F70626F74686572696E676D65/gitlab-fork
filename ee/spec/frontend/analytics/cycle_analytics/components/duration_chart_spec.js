import { GlIcon } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import {
  DURATION_STAGE_TIME_DESCRIPTION,
  DURATION_TOTAL_TIME_DESCRIPTION,
  DURATION_STAGE_TIME_NO_DATA,
  DURATION_TOTAL_TIME_NO_DATA,
} from 'ee/analytics/cycle_analytics/constants';
import DurationChart from 'ee/analytics/cycle_analytics/components/duration_chart.vue';
import Scatterplot from 'ee/analytics/shared/components/scatterplot.vue';
import ChartSkeletonLoader from '~/vue_shared/components/resizable_chart/skeleton_loader.vue';
import { allowedStages as stages, durationChartPlottableData as durationData } from '../mock_data';

Vue.use(Vuex);

const actionSpies = {
  fetchDurationData: jest.fn(),
};

const fakeStore = ({ initialGetters, initialState, rootGetters, rootState }) =>
  new Vuex.Store({
    state: {
      ...rootState,
    },
    getters: {
      isOverviewStageSelected: () => true,
      ...rootGetters,
    },
    modules: {
      durationChart: {
        namespaced: true,
        getters: {
          durationChartPlottableData: () => durationData,
          ...initialGetters,
        },
        state: {
          isLoading: false,
          ...initialState,
        },
        actions: actionSpies,
      },
    },
  });

function createComponent({
  stubs = {},
  initialState = {},
  initialGetters = {},
  rootGetters = {},
  rootState = {},
  props = {},
} = {}) {
  return shallowMount(DurationChart, {
    store: fakeStore({ initialState, initialGetters, rootGetters, rootState }),
    propsData: {
      stages,
      ...props,
    },
    stubs: {
      ChartSkeletonLoader: true,
      Scatterplot: true,
      ...stubs,
    },
  });
}

describe('DurationChart', () => {
  let wrapper;

  const findContainer = (_wrapper) => _wrapper.find('[data-testid="vsa-duration-chart"]');
  const findChartDescription = (_wrapper) => _wrapper.findComponent(GlIcon);
  const findScatterPlot = (_wrapper) => _wrapper.findComponent(Scatterplot);
  const findLoader = (_wrapper) => _wrapper.findComponent(ChartSkeletonLoader);

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('with the overiew stage selected', () => {
    beforeEach(() => {
      wrapper = createComponent({});
    });

    it('renders the duration chart', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders the scatter plot', () => {
      expect(findScatterPlot(wrapper).exists()).toBe(true);
    });

    it('renders the chart description', () => {
      expect(findChartDescription(wrapper).attributes('title')).toBe(
        DURATION_TOTAL_TIME_DESCRIPTION,
      );
    });

    describe('with no chart data', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialGetters: {
            durationChartPlottableData: () => [],
          },
        });
      });

      it('renders the no data available message', () => {
        expect(findContainer(wrapper).text()).toContain(DURATION_TOTAL_TIME_NO_DATA);
      });
    });
  });

  describe('with a value stream stage selected', () => {
    const [selectedStage] = stages;

    beforeEach(() => {
      wrapper = createComponent({
        rootState: {
          selectedStage,
        },
        rootGetters: {
          isOverviewStageSelected: () => false,
        },
      });
    });

    it('renders the scatter plot', () => {
      expect(findScatterPlot(wrapper).exists()).toBe(true);
    });

    it('renders the stage title', () => {
      expect(wrapper.text()).toContain(`Stage time: ${selectedStage.title}`);
    });

    it('sets the scatter plot data', () => {
      expect(findScatterPlot(wrapper).props('scatterData')).toBe(durationData);
    });

    it('sets the median line data', () => {
      expect(findScatterPlot(wrapper).props('medianLineData')).toBe(durationData);
    });

    it('renders the chart description', () => {
      expect(findChartDescription(wrapper).attributes('title')).toBe(
        DURATION_STAGE_TIME_DESCRIPTION,
      );
    });

    describe('with no chart data', () => {
      beforeEach(() => {
        wrapper = createComponent({
          initialGetters: {
            durationChartPlottableData: () => [],
          },
          rootState: {
            selectedStage,
          },
          rootGetters: {
            isOverviewStageSelected: () => false,
          },
        });
      });

      it('renders the no data available message', () => {
        expect(findContainer(wrapper).text()).toContain(DURATION_STAGE_TIME_NO_DATA);
      });
    });
  });

  describe('when isLoading=true', () => {
    beforeEach(() => {
      wrapper = createComponent({ initialState: { isLoading: true } });
    });

    it('renders a loader', () => {
      expect(findLoader(wrapper).exists()).toBe(true);
    });
  });
});
