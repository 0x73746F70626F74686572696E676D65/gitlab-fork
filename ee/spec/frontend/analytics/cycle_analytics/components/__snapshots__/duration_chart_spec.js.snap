// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationChart with the overiew stage selected renders the duration chart 1`] = `
<div
  class="gl-display-flex gl-flex-direction-column"
  data-testid="vsa-duration-chart"
>
  <h4
    class="gl-mt-0"
  >
    
    Total timeÂ 
    <gl-icon-stub
      name="information-o"
      size="16"
      title="The total time items spent across each value stream stage. Data limited to items completed within this date range."
    />
  </h4>
   
  <scatterplot-stub
    medianlinedata="2019-01-01,17,2019-01-01,2019-01-02,40,2019-01-02"
    medianlineoptions="[object Object]"
    scatterdata="2019-01-01,17,2019-01-01,2019-01-02,40,2019-01-02"
    tooltipdateformat="mmm d, yyyy"
    xaxistitle="Date"
    yaxistitle="Average time to completion"
  />
</div>
`;
